src/NuGet.Core/NuGet.Common/InProcLock.cs

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Common
+{
+    internal class InProcLock : IDisposable
+    {
+        private Dictionary<string, LockState> _locks;
+        private SemaphoreSlim _dictionaryLock;
+
+        internal InProcLock()
+        {
+            _locks = new Dictionary<string, LockState>();
```

Comment:
From [`ConcurrentDictionary.AddOrUpdate` docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=netframework-4.8)

> If you call AddOrUpdate simultaneously on different threads, addValueFactory may be called multiple times, but its key/value pair might not be added to the dictionary for every call.

This doesn't fill me with confidence that both threads will get the same `LockState` value.

Reply:
Both threads will definitely get the same `LockState` value; otherwise, the class would be broken.  What the documentation is saying is that the factory method may be called multiple times and some of those values may be discarded.  The first value will be added to the dictionary and returned.  The second value will be discarded and the first value will be returned.

Reply:
I started trying to move to a `ConcurrentDictionary`, but I think it introduces a race condition if there's only 1 thread with a lock on a particular key, and it exits at the same time as a new thread tries to enter. Without using a class lock as I am now, I don't see how to guarantee that when we remove an item from the dictionary that it's really unused.

`Cleanup` needs to do 3 things: release the semaphore, decrement the counter, remove the key from the dictionary if required. No matter which order I consider doing these 3 things, I see a potential timing issue if a different thread calls `GetOrCreate` with the same key.

I think we could guarantee it with a `ReaderWriterLock`, but I'm not sure it would be more readable.